Source/DafnyDriver/CSVTestLogger.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+
+namespace Microsoft.Dafny {
+  public class CSVTestLogger : ITestLoggerWithParameters {
+
+    private readonly ConcurrentBag<TestResult> results = new();
+    private TextWriter writer;
+
+    public void Initialize(TestLoggerEvents events, string testRunDirectory) {
+    }
+
+    public void Initialize(TestLoggerEvents events, Dictionary<string, string> parameters) {
+      events.TestResult += TestResultHandler;
+      events.TestRunComplete += TestRunCompleteHandler;
+      if (parameters.TryGetValue("LogFileName", out string filename)) {
+        writer = filename == "-" ? Console.Out : new StreamWriter(filename);
```

Comment:
Is there a way to avoid the `Console.Out` reference? Seems like both it and the filename have some potential concurrency issues.

How does this `ITestLoggerWithParameters` interface work? Is that something we need to use? It seems a little restrictive.

Reply:
Ah, it looks like retrying if an `IOException` is thrown when opening with `CreateNew` will do the trick.

Reply:
Could you say something more about how `ITestLoggerWithParameters` is called? How is the `parameters` parameter filled in?

Reply:
The BoogieXmlConverter class creates either a CSV or TRX logger and initializes them with parameters based on command-line options, using the code [here](https://github.com/atomb/dafny/blob/csv-logger/Source/DafnyDriver/BoogieXmlConvertor.cs#L34), which is called from [here](https://github.com/atomb/dafny/blob/csv-logger/Source/DafnyDriver/DafnyDriver.cs#L70).

Reply:
There an example of providing those parameters at the command line here FYI: https://github.com/dafny-lang/dafny/blob/master/Test/logger/VerificationLogger.dfy#L1

The `BoogieXmlConverter` class is really translating the data in the Boogie xml file into events using the VSTest `TestResult` class, and then the TRX and/or CSV loggers have event handlers to process those events. In the near future we want to move away from relying on the `/xml` Boogie feature and instead have Boogie raise verification events in a more incremental way for clients. At that point we can translate the Boogie events directly to `TestResult` events, but the loggers can work exactly as they do now (although hence caring about concurrent access, even those that's not possible yet).

We're really borrowing the whole .NET VSTest core model here, which I assert makes sense given that verification tasks are not that different from tests in a development workflow. The core model supports adding extra properties nicely: they won't show up in the TRX file format (which is less extensible), but can easily be added to the CSV output.

Reply:
Alright, thanks :-) PR LGTM